ccmbioinfo/lfs
# CARDS-1344: As a patient-type user, I can access a specific url to start filling out a sequence of questionnaires

### Included in this PR:
* [CARDS-1344](https://phenotips.atlassian.net/browse/CARDS-1344): _As a patient-type user, I can access a specific url to start filling out a sequence of questionnaires_ - 5/6 open sub-tasks
* [CARDS-1358](https://phenotips.atlassian.net/browse/CARDS-1358): _Create a mock patient authentication component_. - a patient identification screen and appropriate patient/visit subject creation. No actual authentication of a patient-type user.
* [CARDS-1349](https://phenotips.atlassian.net/browse/CARDS-1349): _Sync patient-type users and patient subjects_ - patient information forms are created or updated with the demographics data returned after patient identification.

### **NOT** included in this PR:
* [CARDS-1339](https://phenotips.atlassian.net/browse/CARDS-1339): _Patient users should only be able to access the PROMs ui_ (subtask of [CARDS-1344](https://phenotips.atlassian.net/browse/CARDS-1344)) which is dependent on [CARDS-1338](https://phenotips.atlassian.net/browse/CARDS-1338): _Create the patient user type_ - not yet implemented.
* Patient-type user authentication - [CARDS-1337](https://phenotips.atlassian.net/browse/CARDS-1337)
* Final questionnaire structure ([CARDS-1355](https://phenotips.atlassian.net/browse/CARDS-1355)) and access rights on specific questions / sections - ([CARDS-1346](https://phenotips.atlassian.net/browse/CARDS-1346))

### Testing instructions
* Start cards in the `proms` run mode
* Navigate to `/Proms.html/Cardio`
* Log in
* Follow the mock patient identification screen instructions; the following test "patients" are currently recognized by the mock authenticator:
```
[
  {
    last_name: "Addison",
    first_name: "John",
    date_of_birth: "1970-01-01",
    mrn: "1234567",
    health_card: "1234567890AB",
    visit_number: "11"
  }, {
    last_name: "Bennet",
    first_name: "Mary",
    date_of_birth: "1971-02-27",
    mrn: "2345678",
    health_card: "2345678901CD",
    visit_number: "22"
  }, {
    last_name: "Coleman",
    first_name: "Paul",
    date_of_birth: "1972-03-31",
    mrn: "3456789",
    health_card: "3456789012EF",
    visit_number: "33"
  }
];
```
* Also navigate to `/Proms.html` and `/Proms.html/NonExistingProm` to test what happens when the patient goes to the wrong Proms link
* Check that after patient identification there's one Patient Information form for that patient subject, that gets created / updated with the patient data
* Leave some mandatory questions unanswered and check that you're warned about them when you reach the end
* Interrupt answering the questionnaires, navigate away from the page, then return and check that you can pick up where you left off after patient re-identification
* **Note 1: The patient identification functionality is only a mock-up included here to support the survey workflow. There is currently no patient-type user and the patient is not truly authenticated.** Patient authentication will be covered separately by CARDS-1337 and CARDS-1338.
* **Note 2: The questionnaires are not in their final form**, and they will be improved as part of CARDS-1355 (Reviiew PROMs questionnaire) and CARDS-1346 (Configure permissions to prevent patient users from accessing the computed interpretation of their answers). **Testing should focus on the workflow, not on the questionnaire content.** Issues with the questionnaires should be reported as comments on the jira task CARDS-1355.

### Rreviewing instructions
* Reviewing commit by commit rather than file by file is recommended
* The `MockPatientIdentification` is a quick mock-up  included here to support the survey workflow. There is currently no patient-type user and the patient is not truly authenticated. Patient authentication will be covered separately by CARDS-1337 and CARDS-1338, and MockPatientIdentification` will be renamed and modified to support it at that time.
